File: ndai/api/schemas/messaging.py
"""E2E encrypted messaging request/response schemas."""

from pydantic import BaseModel, Field


# --- Prekey upload ---


class OTPKUpload(BaseModel):
    pub: str = Field(max_length=64)
    index: int


class PrekeyBundleUpload(BaseModel):
    identity_x25519_pub: str = Field(max_length=64)
    signed_prekey_pub: str = Field(max_length=64)
    signed_prekey_sig: str = Field(max_length=128)
    signed_prekey_id: int
    one_time_prekeys: list[OTPKUpload]


# --- Prekey fetch response ---


class OTPKResponse(BaseModel):
    pub: str
    index: int


class PrekeyBundleResponse(BaseModel):
    identity_pubkey: str  # Ed25519
    identity_x25519_pub: str
    signed_prekey_pub: str
    signed_prekey_sig: str
    signed_prekey_id: int
    one_time_prekey: OTPKResponse | None  # may be exhausted


class PrekeyStatusResponse(BaseModel):
    remaining_otpks: int
    signed_prekey_age_hours: float


# --- Conversations ---


class ConversationCreate(BaseModel):
    peer_pubkey: str | None = None  # for DM
    agreement_id: str | None = None  # for deal chat


class ConversationResponse(BaseModel):
    id: str
    type: str
    agreement_id: str | None
    participant_a: str
    participant_b: str
    created_at: str
    last_message_at: str | None = None
    unread_count: int = 0


# --- Messages ---


class MessageSend(BaseModel):
    ciphertext: str  # base64
    header: str  # base64
    x3dh_header: str | None = None  # base64, first message only


class MessageResponse(BaseModel):
    id: str
    conversation_id: str
    sender_pubkey: str
    ciphertext: str
    header: str
    x3dh_header: str | None
    message_index: int
    created_at: str

File: frontend-zk/src/api/messaging.ts
/**
 * Messaging API client.
 */
import { get, post } from "./client";

// ─── Prekeys ───────────────────────────────────────────────────────────

export interface PrekeyBundleUpload {
  identity_x25519_pub: string;
  signed_prekey_pub: string;
  signed_prekey_sig: string;
  signed_prekey_id: number;
  one_time_prekeys: { pub: string; index: number }[];
}

export interface PrekeyBundleResponse {
  identity_pubkey: string;
  identity_x25519_pub: string;
  signed_prekey_pub: string;
  signed_prekey_sig: string;
  signed_prekey_id: number;
  one_time_prekey: { pub: string; index: number } | null;
}

export interface PrekeyStatus {
  remaining_otpks: number;
  signed_prekey_age_hours: number;
}

export function uploadPrekeys(bundle: PrekeyBundleUpload) {
  return post("/messaging/prekeys", bundle);
}

export function fetchPeerPrekeys(pubkey: string): Promise<PrekeyBundleResponse> {
  return get<PrekeyBundleResponse>(`/messaging/prekeys/${pubkey}`);
}

export function getPrekeyStatus(): Promise<PrekeyStatus> {
  return get<PrekeyStatus>("/messaging/prekeys/status");
}

// ─── Conversations ─────────────────────────────────────────────────────

export interface ConversationResponse {
  id: string;
  type: string;
  agreement_id: string | null;
  participant_a: string;
  participant_b: string;
  created_at: string;
  last_message_at: string | null;
  unread_count: number;
}

export function createConversation(body: { peer_pubkey?: string; agreement_id?: string }): Promise<ConversationResponse> {
  return post<ConversationResponse>("/messaging/conversations", body);
}

export function listConversations(): Promise<ConversationResponse[]> {
  return get<ConversationResponse[]>("/messaging/conversations");
}

// ─── Messages ──────────────────────────────────────────────────────────

export interface MessageResponse {
  id: string;
  conversation_id: string;
  sender_pubkey: string;
  ciphertext: string;
  header: string;
  x3dh_header: string | null;
  message_index: number;
  created_at: string;
}

export function getMessages(conversationId: string, before?: string, limit = 50): Promise<MessageResponse[]> {
  let url = `/messaging/conversations/${conversationId}/messages?limit=${limit}`;
  if (before) url += `&before=${encodeURIComponent(before)}`;
  return get<MessageResponse[]>(url);
}

export function sendMessage(conversationId: string, body: {
  ciphertext: string;
  header: string;
  x3dh_header?: string | null;
}): Promise<MessageResponse> {
  return post<MessageResponse>(`/messaging/conversations/${conversationId}/messages`, body);
}
